From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: CUB3D <callumthom11@gmail.com>
Date: Wed, 5 Aug 2026 18:05:04 +0000
Subject: [PATCH] 9490: Hexagon: Add missing relocation formats + fix existing
 one

Add missing relocation formats, fix exsting one

Fix duplex

Formatting, use unsigned shifts for divides

Fix HEXAGON_B32_PCREL_X shift on upper half of immext

Start refactoring to masks

Convert more relocations to use masks
---
 .../Hexagon_ElfRelocationHandler.java         | 395 ++++++++++++++++--
 1 file changed, 362 insertions(+), 33 deletions(-)

diff --git a/Ghidra/Processors/Hexagon/src/main/java/ghidra/app/util/bin/format/elf/relocation/Hexagon_ElfRelocationHandler.java b/Ghidra/Processors/Hexagon/src/main/java/ghidra/app/util/bin/format/elf/relocation/Hexagon_ElfRelocationHandler.java
index 6da16e9e51..db43eaf600 100644
--- a/Ghidra/Processors/Hexagon/src/main/java/ghidra/app/util/bin/format/elf/relocation/Hexagon_ElfRelocationHandler.java
+++ b/Ghidra/Processors/Hexagon/src/main/java/ghidra/app/util/bin/format/elf/relocation/Hexagon_ElfRelocationHandler.java
@@ -15,6 +15,9 @@
  */
 package ghidra.app.util.bin.format.elf.relocation;
 
+import java.util.Optional;
+import java.util.stream.Stream;
+
 import ghidra.app.util.bin.format.elf.*;
 import ghidra.app.util.importer.MessageLog;
 import ghidra.program.model.address.Address;
@@ -44,6 +47,107 @@ public class Hexagon_ElfRelocationHandler
 		return Hexagon_ElfRelocationType.R_HEXAGON_RELATIVE.typeId;
 	}
 
+	/*
+	 * Is the given Hexagon sub-instruction a Duplex instruction
+	 */
+	private boolean isDuplex(long ins) {
+		return ((ins >>> 14) & 0b11) == 0b00;
+	}
+
+	/*
+	 * Spread the bits of a value to be relocated based on a mask representing the positions containing the bits of that value.
+	 */
+	private int applyMask(int mask, int val) {
+		int res = 0;
+		int off = 0;
+		for(int i = 0; i < 32; i++) {
+			int valBit = (val >> off) & 1;
+			// If mask[n] is set, fill with next bit from val
+			if (((mask >> i) & 1) != 0) {
+				res |= (valBit << i);
+				off++;
+			}
+
+		}
+		return res;
+	}
+
+	/*
+	 * Find the mask for R_HEX_8_X type relocations
+	 */
+	private int findMaskR8(int ins) {
+		if (isDuplex(ins)) {
+			return 0x03F00000;
+		}
+		switch (ins & 0xFF000000) {
+		case 0xDE000000:
+			return 0x00E020E8;
+		case 0x3C000000:
+			return 0x0000207F;
+		default:
+			return 0x00001FE0;
+		}
+	}
+
+	/*
+	 * Find the mask for R_HEX_11_X type relocations
+	 */
+	private int findMaskR11(int ins) {
+		if (isDuplex(ins)) {
+			return 0x03F00000;
+		}
+		switch (ins & 0xFF000000) {
+		case 0xA1000000:
+			return 0x060020FF;
+		default:
+			return 0x06003FE0;
+		}
+	}
+
+	private record InstructionMask(int compareMask, int valueMask) {}
+
+	private InstructionMask r6Masks[] = new InstructionMask[] {
+		new InstructionMask(0x38000000, 0x0000201f),
+		new InstructionMask(0x39000000, 0x0000201f),
+		new InstructionMask(0x3e000000, 0x00001f80),
+		new InstructionMask(0x3f000000, 0x00001f80),
+		new InstructionMask(0x40000000, 0x000020f8),
+		new InstructionMask(0x41000000, 0x000007e0),
+		new InstructionMask(0x42000000, 0x000020f8),
+		new InstructionMask(0x43000000, 0x000007e0),
+		new InstructionMask(0x44000000, 0x000020f8),
+		new InstructionMask(0x45000000, 0x000007e0),
+		new InstructionMask(0x46000000, 0x000020f8),
+		new InstructionMask(0x47000000, 0x000007e0),
+		new InstructionMask(0x6a000000, 0x00001f80),
+		new InstructionMask(0x7c000000, 0x001f2000),
+		new InstructionMask(0x9a000000, 0x00000f60),
+		new InstructionMask(0x9b000000, 0x00000f60),
+		new InstructionMask(0x9c000000, 0x00000f60),
+		new InstructionMask(0x9d000000, 0x00000f60),
+		new InstructionMask(0x9f000000, 0x001f0100),
+		new InstructionMask(0xab000000, 0x0000003f),
+		new InstructionMask(0xad000000, 0x0000003f),
+		new InstructionMask(0xaf000000, 0x00030078),
+		new InstructionMask(0xd7000000, 0x006020e0),
+		new InstructionMask(0xd8000000, 0x006020e0),
+		new InstructionMask(0xdb000000, 0x006020e0),
+		new InstructionMask(0xdf000000, 0x006020e0),
+	};
+
+	/*
+	 * Find the mask for R_HEX_6_X type relocations
+	 */
+	private Optional<Integer> findMaskR6(int ins) {
+		if (isDuplex(ins)) {
+			return Optional.of(0x03F00000);
+		}
+		return Stream.of(r6Masks)
+			.filter(i -> i.compareMask() == (ins & 0xFF000000))
+			.map(i -> i.valueMask())
+			.findAny();
+	}
+
 	@Override
 	protected RelocationResult relocate(ElfRelocationContext<?> elfRelocationContext,
 			ElfRelocation relocation, Hexagon_ElfRelocationType type, Address relocationAddress,
@@ -100,8 +204,6 @@ public class Hexagon_ElfRelocationHandler
 				memValue |= 0x01ff0000 & (dist << 3);
 				memory.setInt(relocationAddress, memValue);
 				break;
-
-//				break;
 //			case R_HEXAGON_B15_PCREL:
 //				break;
 //			case R_HEXAGON_B7_PCREL:
@@ -136,14 +238,114 @@ public class Hexagon_ElfRelocationHandler
 				byteLength = 1;
 				break;
 
-//			case R_HEXAGON_GPREL16_0:
-//				break;
-//			case R_HEXAGON_GPREL16_1:
-//				break;
-//			case R_HEXAGON_GPREL16_2:
-//				break;
-//			case R_HEXAGON_GPREL16_3:
-//				break;
+			case R_HEXAGON_GPREL16_0:
+				if(
+					// Rd=memb(gp+#u16:0)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1000)) ||
+					// Rd=memub(gp+#u16:0)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1001))
+				) {
+					memValue &= ~0x61F3FE0;
+					memValue |= (value & 0x1FF) << 5;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				} else if (
+					// memb(gp+#u16:0)=Rt
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0000)) ||
+					// memb(gp+#u16:0)=Nt.new
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0101) && (((memValue >> 11) & 0b11) == 0b00))
+				) {
+					memValue &= ~0x61f20ff;
+					memValue |= (value & 0xFF);
+					memValue |= ((value >> 8) & 1) << 13;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				}
+				else {
+					markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName, log);
+					return RelocationResult.UNSUPPORTED;
+				}
+				break;
+			case R_HEXAGON_GPREL16_1:
+				value >>>= 1;
+				if(
+					// Rd=memh(gp+#u16:1)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1010)) ||
+					// Rd=memuh(gp+#u16:1)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1011))
+				) {
+					memValue &= ~0x61F3FE0;
+					memValue |= (value & 0x1FF) << 5;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				} else if (
+					// memh(gp+#u16:1)=Rt
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0010)) ||
+					// memh(gp+#u16:1)=Rt.H
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0011)) ||
+					// memh(gp+#u16:1)=Nt.new
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0101) && (((memValue >> 11) & 0b11) == 0b01))
+				) {
+					memValue &= ~0x61f20ff;
+					memValue |= (value & 0xFF);
+					memValue |= ((value >> 8) & 1) << 13;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				}
+				else {
+					markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName, log);
+					return RelocationResult.UNSUPPORTED;
+				}
+				break;
+			case R_HEXAGON_GPREL16_2:
+				value >>>= 2;
+				if(
+					// Rd=memw(gp+#u16:2)
+					(((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1100)
+				) {
+					memValue &= ~0x61F3FE0;
+					memValue |= (value & 0x1FF) << 5;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				} else if (
+					// memw(gp+#u16:2)=Rt
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0100)) ||
+					// memw(gp+#u16:2)=Nt.new
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0101) && (((memValue >> 11) & 0b11) == 0b10))
+				) {
+					memValue &= ~0x61f20ff;
+					memValue |= (value & 0xFF);
+					memValue |= ((value >> 8) & 1) << 13;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				}
+				else {
+					markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName, log);
+					return RelocationResult.UNSUPPORTED;
+				}
+				break;
+			case R_HEXAGON_GPREL16_3:
+				value >>>= 3;
+				if(
+					// Rdd=memd(gp+#u16:3)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1110))
+				) {
+					memValue &= ~0x61F3FE0;
+					memValue |= (value & 0x1FF) << 5;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				} else {
+					markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName, log);
+					return RelocationResult.UNSUPPORTED;
+				}
+				break;
 //			case R_HEXAGON_HL16:
 //				break;
 //			case R_HEXAGON_B13_PCREL:
@@ -157,38 +359,165 @@ public class Hexagon_ElfRelocationHandler
 				memValue &= ~0x0fff3fff;
 				memValue |= dist & 0x3fff;
 				dist = dist >>> 14;
-				memValue |= (dist << 20);
+				memValue |= ((dist & 0xFFF) << 16);
 				memory.setInt(relocationAddress, memValue);
 				byteLength = 4;
 				break;
-//			case R_HEXAGON_32_6_X:
-//				break;
+			case R_HEXAGON_32_6_X:
+				// This relocation is used to handle extended immediates
+				int c = (value >>> 6);
+				memValue &= ~0x0fff3fff;
+				memValue |= (c & 0x3FFF);
+				memValue |= ((c >> 14) & 0xFFF) << 16;
+				memory.setInt(relocationAddress, memValue);
+				break;
 //			case R_HEXAGON_B22_PCREL_X:
 //				break;
-//			case R_HEXAGON_B15_PCREL_X:
-//				break;
+			case R_HEXAGON_B15_PCREL_X:
+				dist = (int) (Integer.toUnsignedLong(value) - Integer.toUnsignedLong((int) offset)) & 0x3f;
+				dist >>>= 6;
+				if(
+					// if(pu) call #r15:2
+					((((memValue >> 24) & 0b11111111) == 0b01011101) && (((memValue >> 21) & 1) == 0) && (((memValue >> 11) & 1) == 0)) ||
+					// if(!pu) call #r15:2
+					((((memValue >> 24) & 0b11111111) == 0b01011101) && (((memValue >> 21) & 1) == 1) && (((memValue >> 11) & 1) == 0))
+				) {
+					memValue &= ~0xDF20FE;
+					memValue |= (dist & 0x7F) << 1;
+					memValue |= ((dist >> 7) & 1) << 13;
+					memValue |= ((dist >> 8) & 0x1F) << 16;
+					memValue |= ((dist >> 13) & 0x3) << 22;
+					memory.setInt(relocationAddress, memValue);
+				} else {
+					markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName, log);
+					return RelocationResult.UNSUPPORTED;
+				}
+				break;
 //			case R_HEXAGON_B13_PCREL_X:
 //				break;
-//			case R_HEXAGON_B9_PCREL_X:
-//				break;
+			case R_HEXAGON_B9_PCREL_X:
+				dist = (int) (Integer.toUnsignedLong(value) - Integer.toUnsignedLong((int) offset)) & 0x3F;
+				if(
+					// Rd=u6 ; jump #r9:2
+					(((memValue >> 24) & 0b11111111) == 0b00010110) ||
+					// Rd=Rs ; jump #r9:2
+					(((memValue >> 24) & 0b11111111) == 0b00010111)
+				) {
+					memValue &= ~0x3000FE;
+					memValue |= (dist & 0x7F) << 1;
+					memValue |= ((dist >> 7) & 0x3) << 20;
+					memory.setInt(relocationAddress, memValue);
+				} else {
+					markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName, log);
+					return RelocationResult.UNSUPPORTED;
+				}
+				break;
 //			case R_HEXAGON_B7_PCREL_X:
 //				break;
-//			case R_HEXAGON_16_X:
-//				break;
-//			case R_HEXAGON_12_X:
-//				break;
-//			case R_HEXAGON_11_X:
-//				break;
-//			case R_HEXAGON_10_X:
-//				break;
-//			case R_HEXAGON_9_X:
-//				break;
-//			case R_HEXAGON_8_X:
-//				break;
-//			case R_HEXAGON_7_X:
-//				break;
-//			case R_HEXAGON_6_X:
-//				break;
+			case R_HEXAGON_16_X:
+				//Rd=#s16
+				if(((memValue >> 24) & 0b11111111) == 0b01111000) {
+					memValue &= ~0xDF3FF0;
+					memValue |= (value & 0x1FF) << 5;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 22;
+					memory.setInt(relocationAddress, memValue);
+				} else if (
+					// Rd=memb(gp+#u16:0)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1000)) ||
+					// Rd=memh(gp+#u16:1)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1010)) ||
+					// Rd=memw(gp+#u16:2)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1100)) ||
+					// Rdd=memd(gp+#u16:3)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1110)) ||
+					// Rd=memub(gp+#u16:0)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1001)) ||
+					// Rd=memuh(gp+#u16:1)
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b1011))
+				) {
+					memValue &= ~0x61F3F70;
+					memValue |= (value & 0x1FF) << 5;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				} else if (
+					// memb(gp+#u16:0)=Rt
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0000)) ||
+					// memh(gp+#u16:1)=Rt
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0010)) ||
+					// memh(gp+#u16:1)=Rt.h
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0011)) ||
+					// memw(gp+#u16:2)=Rt
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0100)) ||
+					// memd(gp+#u16:3)=Rtt
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0110))
+				) {
+					memValue &= ~0x61f20ff;
+					memValue |= (value & 0xFF);
+					memValue |= ((value >> 8) & 1) << 13;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				} else if (
+					// memb(gp+#u16:0)=Nt.new
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0101) && (((memValue >> 11) & 0b11) == 0b00)) ||
+					// memh(gp+#u16:1)=Nt.new
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0101) && (((memValue >> 11) & 0b11) == 0b01)) ||
+					// memw(gp+#u16:2)=Nt.new
+					((((memValue >> 27) & 0b11111) == 0b01001) && (((memValue >> 21) & 0b1111) == 0b0101) && (((memValue >> 11) & 0b11) == 0b10))
+				) {
+					memValue &= ~0x61f20ff;
+					memValue |= (value & 0xFF);
+					memValue |= ((value >> 8) & 1) << 13;
+					memValue |= ((value >> 9) & 0x1F) << 16;
+					memValue |= ((value >> 14) & 0x3) << 25;
+					memory.setInt(relocationAddress, memValue);
+				} else {
+					markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName, log);
+					return RelocationResult.UNSUPPORTED;
+				}
+				break;
+			case R_HEXAGON_12_X:
+				int mask = 0x000F1FE0;
+				memValue |= applyMask(mask, value & 0x3F);
+				memory.setInt(relocationAddress, memValue);
+				break;
+			case R_HEXAGON_11_X:
+				mask = findMaskR11(memValue);
+				memValue |= applyMask(mask, value & 0x3F);
+				memory.setInt(relocationAddress, memValue);
+				break;
+			case R_HEXAGON_10_X:
+				mask = 0x0020EFE0;
+				memValue |= applyMask(mask, value & 0x3F);
+				memory.setInt(relocationAddress, memValue);
+				break;
+			case R_HEXAGON_9_X:
+				mask = 0x00003FE0;
+				memValue |= applyMask(mask, value  & 0x3F);
+				memory.setInt(relocationAddress, memValue);
+				break;
+			case R_HEXAGON_8_X:
+				mask = findMaskR8(memValue);
+				memValue |= applyMask(mask, value & 0x3F);
+				memory.setInt(relocationAddress, memValue);
+				break;
+			case R_HEXAGON_7_X:
+				mask = 0x00000FE0;
+				memValue |= applyMask(mask, value & 0x3F);
+				memory.setInt(relocationAddress, memValue);
+				break;
+			case R_HEXAGON_6_X:
+				Optional<Integer> r6Mask = findMaskR6(memValue);
+				if(r6Mask.isEmpty()) {
+					markAsUnhandled(program, relocationAddress, type, symbolIndex, symbolName, log);
+					return RelocationResult.UNSUPPORTED;
+				}
+				mask = r6Mask.get();
+				memValue |= applyMask(mask, value & 0x3F);
+				memory.setInt(relocationAddress, memValue);
+				break;
 
 			case R_HEXAGON_32_PCREL:
 				dist = (int) (Integer.toUnsignedLong(value) - Integer.toUnsignedLong((int) offset));
-- 
2.45.1

